homeassistant/components/hunterdouglas_powerview/shade_data.py

```diff
@@ -1,59 +1,25 @@
 """Shade data for the Hunter Douglas PowerView integration."""
 from __future__ import annotations
 
-from collections.abc import Iterable
-from dataclasses import dataclass
 import logging
 from typing import Any
 
-from aiopvapi.helpers.constants import (
-    ATTR_ID,
-    ATTR_POSITION1,
-    ATTR_POSITION2,
-    ATTR_POSITION_DATA,
-    ATTR_POSKIND1,
-    ATTR_POSKIND2,
-    ATTR_SHADE,
-)
-from aiopvapi.resources.shade import MIN_POSITION
-
-from .const import POS_KIND_PRIMARY, POS_KIND_SECONDARY, POS_KIND_VANE
-from .util import async_map_data_by_id
+from aiopvapi.resources.model import PowerviewData
+from aiopvapi.resources.shade import BaseShade, ShadePosition
 
-POSITIONS = ((ATTR_POSITION1, ATTR_POSKIND1), (ATTR_POSITION2, ATTR_POSKIND2))
+from .util import async_map_data_by_id
 
 _LOGGER = logging.getLogger(__name__)
 
 
-@dataclass
-class PowerviewShadeMove:
-    """Request to move a powerview shade."""
-
-    # The positions to request on the hub
-    request: dict[str, int]
-
-    # The positions that will also change
-    # as a result of the request that the
-    # hub will not send back
-    new_positions: dict[int, int]
-
-
-@dataclass
-class PowerviewShadePositions:
-    """Positions for a powerview shade."""
-
-    primary: int = MIN_POSITION
-    secondary: int = MIN_POSITION
-    vane: int = MIN_POSITION
-
-
 class PowerviewShadeData:
```

Comment:
Why isn't this part of the 3rd party library?

Reply:
Home Assistant keeps a copy of the data as the ip stream does not return positional data for all movement types. This is a failure in the design by Hunter Douglas. Their newer Gen 3 has fixed this but we need this to handle gen 2 data inconsistencies

Reply:
Ok, but it's there a reason why it can't be handled in the 3rd party library?

Reply:
Guess I'm not sure what the issue is here - I've stripped a lot of code out of HA and moved to the upstream already - most of the code in this class is being accessed by the co-ordinator as we need to send it back to the hub
Ie when moving a TDBU shade current HA architecture only allows one shade to move (top or bottom). We need to send the current position of the opposing shade in the command aswell or the opposite shade will close. Same for some tilt shades

Reply:
The issue is that it looks like we're handling details of the api. Normally api details should be handled by the 3rd party library.

If this code is just a consequence of the Home Assistant entity design, it's ok to keep it in Home Assistant. But if it's a problem in the api design and library users would generally benefit from this code, the code should move to the 3rd party library.

Reply:
Yea this is how HA interacts with the API - the API expects to receive certain values for a move request and if not received the shade defaults to 0

HA stores the information to know what to send in the move request

Reply:
Is this not useful for other library users?

Reply:
The api expects the two values, it doesnt dictate they need to exist.
This is just how the hub acts if they are ommitted and could also be desirable by some users.
The library treats it the same way - if you provide a value it sends it on, if not then we do not ASSUME that you wanted us to

Reply:
We should move this to the library, but lets not do it right now since its going to be too much to review an architecture change and v3 support at the same time, and this PR is already large.